packages/react-components/react-utilities/src/compose/slot.ts

```diff
@@ -0,0 +1,67 @@
+import { isValidElement } from 'react';
+import type {
+  AsIntrinsicElement,
+  SlotComponent,
+  SlotComponentMetadata,
+  SlotRenderFunction,
+  SlotShorthandValue,
+  UnknownSlotProps,
+} from './types';
+import { SLOT_COMPONENT_METADATA_SYMBOL, SLOT_COMPONENT_TYPEOF_SYMBOL } from './constants';
+import * as React from 'react';
+
+export type SlotOptions<Props extends UnknownSlotProps> = {
+  componentType:
+    | React.ComponentType<Props>
+    | (Props extends AsIntrinsicElement<infer As> ? As : keyof JSX.IntrinsicElements);
+  defaultProps?: Partial<Props>;
+  overrides?: Partial<Props>;
+};
+
+export function slot<Props extends UnknownSlotProps = UnknownSlotProps>(
+  shorthand: Props | SlotShorthandValue | undefined,
+  options: SlotOptions<Props> & { required: true },
+): SlotComponent<Props>;
+export function slot<Props extends UnknownSlotProps = UnknownSlotProps>(
+  shorthand: Props | SlotShorthandValue | null | undefined,
+  options: SlotOptions<Props> & { required?: boolean },
+): SlotComponent<Props> | undefined;
+export function slot<Props extends UnknownSlotProps = UnknownSlotProps>(
+  shorthand: Props | SlotShorthandValue | null | undefined,
+  options: SlotOptions<Props> & { required?: boolean },
+): SlotComponent<Props> | undefined {
+  const { required = false, defaultProps, overrides, componentType } = options;
+
+  if (shorthand === null || (shorthand === undefined && !required)) {
+    return undefined;
+  }
+  const metadata: SlotComponentMetadata<Props> = { componentType };
+
+  const slotComponent = {
+    ...defaultProps,
+    $$typeof: SLOT_COMPONENT_TYPEOF_SYMBOL,
+    [SLOT_COMPONENT_METADATA_SYMBOL]: metadata,
+  } as SlotComponent<Props>;
+
+  if (
+    typeof shorthand === 'string' ||
+    typeof shorthand === 'number' ||
+    Array.isArray(shorthand) ||
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    isValidElement<any>(shorthand)
+  ) {
+    slotComponent.children = shorthand;
+  } else if (typeof shorthand === 'object') {
+    Object.assign(slotComponent, shorthand);
+    if (typeof shorthand.children === 'function') {
+      metadata.renderFunction = shorthand.children as SlotRenderFunction<Props>;
+      slotComponent.children = defaultProps?.children;
+    }
+  }
+
+  return Object.assign(slotComponent, overrides);
```

Comment:
```suggestion
  return slotComponent;
```

Let's remove `overrides` as a concept from this implementation. It had sense in Northstar as factories returned there a React element, but here it returns `props`. With overrides there will be two ways of doing things:

```ts
const a = slot(props.a)
slot.foo = 'foo'

// vs

const a = slot(props.a, { overrides: { foo: 'foo' })
```

What is a correct way? And would it should be done via `overrides`? 🐱

Reply:
Well, yeah. we can remove it for sure. https://github.com/microsoft/fluentui/pull/27753#discussion_r1184749654, this becomes necessary though.